packages/storage-plus/src/indexed_snapshot.rs

```diff
@@ -0,0 +1,677 @@
+// this module requires iterator to be useful at all
+#![cfg(feature = "iterator")]
+
+use cosmwasm_std::{StdError, StdResult, Storage};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use crate::keys::{EmptyPrefix, Prefixer, PrimaryKey};
+use crate::prefix::{Bound, Prefix};
+use crate::snapshot::SnapshotMap;
+use crate::{IndexList, Strategy};
+
+/// IndexedSnapshotMap works like a SnapshotMap but has a secondary index
+/// TODO: remove traits here and make this const fn new
+pub struct IndexedSnapshotMap<'a, K, T, I>
+where
+    K: PrimaryKey<'a>,
+    T: Serialize + DeserializeOwned + Clone,
+    I: IndexList<T>,
+{
+    pk_namespace: &'a [u8],
+    primary: SnapshotMap<'a, K, T>,
```

Comment:
This is a copy of indexed_map.rs with some mods, right?

But how do you make the historical queries? Those special functions of SnapshotMap need to be exposed somehow.

Reply:
I'm now working on using this `IndexedSnapshotMap` to implement `list_members` in `cw4-group`.

Looks good so far, but I'm making some shortcuts:

I'm making the `primary` member of `IndexedSnapshotMap` `pub`, so that I can access the wrapped `SnapshotMap` through it directly. This is not very good, but it's simple, and it's what is being done for accessing the `idx`s.

An alternative would be to wrap the needed methods; or use the `Deref` trait trick to expose them.

Another alternative would be, just create an external, similar `SnapshotMap`s. Sone one is inside the `Indexed*` wrapper, the other is not (and can be `const`), like the current `MEMBERS` is. These are just namespaces in the end.

And, even another alternative, would be to pass a `SnapshotMap` instance as parameter for the `IndexedSnapshotMap` constructor.

More about this later.

Other shortcut is: I'm just re-implementing `list_members`, and the associated `ListMembers`  message, using this new indexed map. Maybe you would prefer having two different messages? i.e. `ListMembers` which is the traditional, unordered functionality, and a `ListMembersByWeight` which implements members ordered by descending weight.

Another option would be to have a message called `TopMembers`, which can implement limits but not pagination, by example.

Good news is that it's working. I'm having some issues with prefixes and ranges, but the overall functionality is already working.

Reply:
> Maybe you would prefer having two different messages? i.e. ListMembers which is the traditional, unordered functionality, and a ListMembersByWeight which implements members ordered by (descending) weight.

I would prefer a new method. I think the existing implementation works well enough. You can gladly add an extra method to `cw4-group` (but not `cw4` spec) to demo this.

Reply:
> I'm making the primary member of IndexedSnapshotMap pub, so that I can access the wrapped SnapshotMap through it directly. This is not very good, but it's simple, and it's what is being done for accessing the idxs.

This is a very good choice for trying it out.

> An alternative to that would be to wrap the needed methods; or use the Deref trait trick to expose them.

`Deref` seems too much of a hack here (this is far from a "smart pointer") and there are not so many methods, so I would just manually add them.

(And on later reflection, I think making this generic over Map and SnapshotMap is more complexity than it is worth and we will end up with something we will have trouble maintaining)

Reply:
See #272 for a (working) impl. We can then discuss some details.

I'll separate it into two different calls, as mentioned above, when finding some time tomorrow.

Reply:
> > I'm making the primary member of IndexedSnapshotMap pub, so that I can access the wrapped SnapshotMap through it directly. This is not very good, but it's simple, and it's what is being done for accessing the idxs.
> 
> This is a very good choice for trying it out.

We can simply do that then.

> 
> > An alternative to that would be to wrap the needed methods; or use the Deref trait trick to expose them.
> 
> `Deref` seems too much of a hack here (this is far from a "smart pointer") and there are not so many methods, so I would just manually add them.
> 
> (And on later reflection, I think making this generic over Map and SnapshotMap is more complexity than it is worth and we will end up with something we will have trouble maintaining)

Agreed.